src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -0,0 +1,349 @@
+/// \brief Envirtonment Sensor used for looking up environment values in our
+/// CSV file.
+class EnvironmentalSensor : public gz::sensors::Sensor
+{
+  /// \brief Node for communication
+  private: gz::transport::Node node;
+
+  /// \brief Publishes sensor data
+  private: gz::transport::Node::Publisher pub;
+
+  ////////////////////////////////////////////////////////////////
+  /// Documentation inherited
+  public: virtual bool Load(const sdf::Sensor &_sdf) override
+  {
+    auto type = gz::sensors::customType(_sdf);
+    if (type.substr(0, strlen(SENSOR_TYPE_PREFIX)) != SENSOR_TYPE_PREFIX ||
+      type.size() <= strlen(SENSOR_TYPE_PREFIX))
+    {
+      gzerr << "Trying to load [" << SENSOR_TYPE_PREFIX
+        << "] sensor, but got type ["
+        << type << "] instead." <<  std::endl;
+      return false;
+    }
+
+    // Load common sensor params
+    gz::sensors::Sensor::Load(_sdf);
+
+    this->pub = this->node.Advertise<gz::msgs::Double>(this->Topic());
+
+    this->field = type.substr(strlen(SENSOR_TYPE_PREFIX));
+
+    gzdbg << "Loaded environmental sensor for " << this->field
+      << " publishing on " << this->Topic() << std::endl;
+
+    return true;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Update the sensor and generate data
+  /// \param[in] _now The current time
+  /// \return True if the update was successful
+  public: virtual bool Update(
+    const std::chrono::steady_clock::duration &_now) override
+  {
+    if (!this->ready) return false;
+
+    if (!this->session.has_value()) return false;
+
+    this->session = this->gridField->frame[this->field].StepTo(
+      this->session.value(), std::chrono::duration<double>(_now).count());
+
+    if (!this->session.has_value()) return false;
+
+    gz::msgs::Double msg;
+    *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
+    auto frame = msg.mutable_header()->add_data();
+    frame->set_key("frame_id");
+    frame->add_value(this->Name());
+    auto data = this->gridField->frame[this->field].LookUp(
+      this->session.value(), this->position);
+    if (!data.has_value())
+      return false;
+    msg.set_data(data.value());
+    // TODO(anyone) Add sensor noise.
+    this->pub.Publish(msg);
+    return true;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Attempts to set a data table.
+  /// \param[in] _data - The data table
+  /// \param[in] _curr_time - The current time.
+  public: void SetDataTable(
+    const components::Environment* _data,
+    const std::chrono::steady_clock::duration &_curr_time)
+  {
+    gzdbg << "Setting new data table\n";
+    auto data = _data->Data();
+    if(!data->frame.Has(this->field))
+    {
+      gzwarn << "Environmental sensor could not find field "
+        << this->field << "\n";
+      ready = false;
+      return;
+    }
+
+    this->gridField = data;
+    this->session = this->gridField->frame[this->field].CreateSession();
```

Comment:
@arjo129 nit^2: why not just keep a reference to the volumetric grid we care about? As opposed to indexing the data frame each and every time.

Reply:
We could do that but that would mean that I don't hold on to the refcount for the std::shared_ptr. So if the data goes out of memory somewhere else this plugin will :skull:

Reply:
Right. We can distribute a `std::shared_ptr` to each and every `EnvironmentalSensor`, OR we could keep a single `std::shared_ptr` in `EnvironmentalSensorSystem` and distribute a non-owning reference (ie. a pointer) to each `EnvironmentalSensor`.